R/binary_beta.R

```diff
@@ -0,0 +1,2567 @@
+  # ----------------------------------------------------------------------------
+  .ph_log_info(
+    "performing principal coordinates analysis",
+    bullets = if (identical(neg_correction, "none")) NULL else
+      paste("using", neg_correction, "correction")
+  )
+
+  # cmdscale requires k in [1, n - 1]
+  k_cmd <- min(n_axes, n - 1L)
+
+  pcoa_fit <- if (identical(neg_correction, "none")) {
+    stats::cmdscale(dist_obj, eig = TRUE, k = k_cmd)
+  } else {
+    vegan::wcmdscale(dist_obj, eig = TRUE, k = k_cmd, add = neg_correction)
+  }
+
+  is_corrected <- !identical(neg_correction, "none")
+  correction_add <- if (is_corrected && !is.null(pcoa_fit$add)) {
+    as.numeric(pcoa_fit$add)
+  } else {
+    NA_real_
+  }
+  correction_note <- if (is_corrected) {
+    sprintf(
+      "Distances modified using %s correction prior to PCoA.",
+      neg_correction
+    )
+  } else {
+    "No negative eigenvalue correction applied."
+  }
+
+  if (is_corrected) {
+    .ph_warn(correction_note)
+  }
+
+  eig_vals <- pcoa_fit$eig
+  if (is.null(eig_vals)) eig_vals <- numeric(0L)
+  eig_vals <- as.numeric(eig_vals)
+
+  coords <- as.matrix(pcoa_fit$points)
+  if (is.null(coords)) {
+    coords <- matrix(0, nrow = n, ncol = 0L)
+  }
+
+  # enforce stable axis and sample naming
+  if (ncol(coords) > 0L) {
+    colnames(coords) <- paste0("PCoA", seq_len(ncol(coords)))
+  }
+  if (nrow(coords) == n) {
+    rownames(coords) <- labels
+  }
+
+  # ----------------------------------------------------------------------------
+  # 3) sample coordinates (first n_axes, or fewer)
+  # ----------------------------------------------------------------------------
+  .ph_log_info("extracting sample coordinates.")
+  k_use <- min(n_axes, ncol(coords))
+  coords_k <- if (k_use > 0L) {
+    coords[, seq_len(k_use), drop = FALSE]
+  } else {
+    matrix(
+      numeric(0L),
+      nrow = n,
+      ncol = 0L,
+      dimnames = list(labels, character(0L))
+    )
+  }
+
+  sample_coords <- tibble::as_tibble(coords_k, rownames = "sample_id")
+
+  # ----------------------------------------------------------------------------
+  # 4) variance explained (based on positive eigenvalues)
+  # ----------------------------------------------------------------------------
+  .ph_log_info("summarizing eigenvalues and variance explained.")
+  pos_eig <- pmax(eig_vals, 0)
+  sum_pos <- sum(pos_eig, na.rm = TRUE)
+  neg_eig <- pmin(eig_vals, 0)
+  sum_neg <- sum(abs(neg_eig), na.rm = TRUE)
+
+  if (sum_pos > 0 && k_use > 0L) {
+    pct_axes <- 100 * pos_eig[seq_len(k_use)] / sum_pos
+    pct_other <- if (length(pos_eig) > k_use) {
+      100 * sum(pos_eig[(k_use + 1L):length(pos_eig)], na.rm = TRUE) / sum_pos
+    } else {
+      0
+    }
+  } else {
+    pct_axes <- rep(NA_real_, k_use)
+    pct_other <- NA_real_
+  }
+
+  names(pct_axes) <- paste0("%PCoA", seq_len(k_use))
+  var_explained <- tibble::as_tibble_row(
+    c(as.list(round(pct_axes, 3)), `%Other` = round(pct_other, 3))
```

Comment:
I wouldn't round - let the user have the information and round it if and when they want